File: backend/src/main/kotlin/net/animeta/backend/serializer/WorkSerializer.kt
package net.animeta.backend.serializer

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.google.common.net.UrlEscapers
import net.animeta.backend.dto.WorkDTO
import net.animeta.backend.dto.WorkLinks
import net.animeta.backend.dto.WorkMetadata
import net.animeta.backend.dto.WorkSchedule
import net.animeta.backend.model.Period
import net.animeta.backend.model.User
import net.animeta.backend.model.Work
import net.animeta.backend.repository.RecordRepository
import net.animeta.backend.service.WorkService
import org.springframework.stereotype.Service
import java.io.IOException
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.temporal.Temporal

@Service
class WorkSerializer(val workService: WorkService,
                     val recordRepository: RecordRepository,
                     val recordSerializer: RecordSerializer,
                     val objectMapper: ObjectMapper) {
    private val defaultTimeZone = ZoneId.of("Asia/Seoul")

    fun serialize(work: Work, viewer: User? = null, full: Boolean = false): WorkDTO {
        val index = work.indexes.firstOrNull()
        return WorkDTO(
                id = work.id!!,
                title = work.title,
                image_url = getImageUrl(work),
                alt_titles = if (full) {
                    work.titleMappings.map { it.title }.filterNot { it == work.title }
                } else {
                    null
                },
                episodes = if (full) workService.getEpisodes(work) else null,
                record_count = index?.record_count ?: recordRepository.countByWork(work),
                rank = index?.rank,
                record = if (viewer != null) {
                    recordRepository.findOneByWorkAndUser(work, viewer)
                            ?.let { recordSerializer.serialize(it) }
                } else {
                    null
                },
                metadata = if (work.metadata != null) {
                    try {
                        serializeMetadata(work, objectMapper.readTree(work.metadata))
                    } catch (e: IOException) {
                        null
                    }
                } else {
                    null
                }
        )
    }

    fun getImageUrl(work: Work): String? {
        if (work.image_filename != null)
            return "https://animeta.net/media/" + work.image_filename
        return null
    }

    fun serializeMetadata(work: Work, item: JsonNode): WorkMetadata {
        var titleNode: JsonNode? = item["title"]
        if (titleNode != null && titleNode.isObject()) {
            titleNode = titleNode["ko"]
        }
        val title = titleNode?.asText() ?: work.title

        val links = WorkLinks(
                website = item["website"]?.asText(),
                namu = item["namu_ref"]?.asText()?.let { namuLink(it) },
                ann_id = item["ann_id"]?.asText()?.let { "http://www.animenewsnetwork.com/encyclopedia/anime.php?id=${it}" }
        )

        val period = item["periods"]?.let { readStringList(it) }
                ?.map { Period.parse(it) }?.firstOrNull()
                ?: Period.now(defaultTimeZone)

        return WorkMetadata(
                title = title,
                links = links,
                studios = item["studio"]?.let { readStringList(it) },
                source = item["source"]?.asText(),
                schedule = mapOf(
                        "jp" to item["schedule"]?.let { getSchedule(it, period) },
                        "kr" to item["schedule_kr"]?.let { getSchedule(it, period) }
                )
        )
    }

    fun namuLink(ref: String): String {
        val page = ref.substringBeforeLast('#')
        val anchor = ref.substringAfterLast('#', "")
        val url = "https://namu.wiki/w/${UrlEscapers.urlPathSegmentEscaper().escape(page)}"
        if (anchor.isNotEmpty()) {
            return "${url}#${UrlEscapers.urlFragmentEscaper().escape(anchor)}"
        } else {
            return url
        }
    }

    fun readStringList(node: JsonNode): List<String> {
        if (node.isTextual) {
            return listOf(node.asText())
        } else {
            return node.map { it.asText() }
        }
    }

    fun getSchedule(node: JsonNode, period: Period): WorkSchedule? {
        var date: Temporal?
        var broadcasts: List<String>?
        if (node.isTextual) {
            date = parseDateTime(node.asText(), period)
            broadcasts = null
        } else if (node.size() == 1) {
            date = null
            broadcasts = readStringList(node[0])
        } else {
            date = parseDateTime(node[0].asText(), period)
            broadcasts = readStringList(node[1])
        }
        return when (date) {
            is LocalDate -> WorkSchedule(
                    date = date.atStartOfDay(defaultTimeZone).toInstant().toEpochMilli(),
                    date_only = true,
                    broadcasts = broadcasts
            )
            is LocalDateTime -> WorkSchedule(
                    date = date.atZone(defaultTimeZone).toInstant().toEpochMilli(),
                    date_only = false,
                    broadcasts = broadcasts
            )
            else -> WorkSchedule(
                    date = null,
                    date_only = null,
                    broadcasts = broadcasts
            )
        }
    }

    private fun parseDateTime(s: String, period: Period): Temporal {
        val parts = s.split(" ")
        if (parts.size == 1) {
            return parseDate(s, period)
        }
        val date = parseDate(parts[0], period)
        val timeParts = parts[1].split(":")
        return date.atTime(timeParts[0].toInt(), timeParts[1].toInt())
    }

    private fun parseDate(s: String, period: Period): LocalDate {
        val parts = s.split("-")
        return if (parts.size == 3) {
            LocalDate.of(parts[0].toInt(), parts[1].toInt(), parts[2].toInt())
        } else {
            LocalDate.of(period.year, parts[0].toInt(), parts[1].toInt())
        }
    }
}

File: backend-ts/src/serializers/work.serializer.ts
import { Injectable } from "@nestjs/common";
import { WorkService } from "src/services/work.service";
import { Repository } from "typeorm";
import { Record } from 'src/entities/record.entity';
import { RecordSerializer } from "./record.serializer";
import { InjectRepository } from "@nestjs/typeorm";
import { Work } from "src/entities/work.entity";
import { User } from "src/entities/user.entity";
import { WorkDTO, WorkLinks, WorkMetadataDTO, WorkSchedule } from "shared/types_generated";
import { RecordService } from "src/services/record.service";
import { Schedule } from "src/entities/work_metadata";
import { Temporal } from "@js-temporal/polyfill";

@Injectable()
export class WorkSerializer {
  private defaultTimeZone = Temporal.TimeZone.from('Asia/Seoul')

  constructor(
    private workService: WorkService,
    private recordService: RecordService,
    @InjectRepository(Record) private recordRepository: Repository<Record>,
    private recordSerializer: RecordSerializer,
  ) {}

  async serialize(
    work: Work,
    viewer: User | null = null,
    full: boolean = false
  ): Promise<WorkDTO> {
    const index = await this.workService.getIndex(work.id)
    const record = viewer ? await this.recordService.findByUserAndWorkId(viewer, work.id) : null
    return {
      id: work.id,
      title: work.title,
      image_url: this.workService.getImageUrl(work),
      image_center_y: work.image_center_y,
      episodes: full ? await this.workService.getEpisodes(work) : null,
      record_count: index?.record_count ?? await this.recordRepository.count({ where: {work_id: work.id} }),
      record: record ? await this.recordSerializer.serialize(record, RecordSerializer.legacyOptions()) : null,
      metadata: this.serializeMetadata(work),

      recommendations: null,
      recommendationScore: 0,
    }
  }

  private serializeMetadata(work: Work): WorkMetadataDTO | null {
    const metadata = work.metadata
    if (!metadata) return null

    const title = metadata.title ?? work.title

    const links: WorkLinks = {
      website: metadata.website ?? null,
      namu: metadata.namuRef ? this.namuLink(metadata.namuRef) : null,
      ann: metadata.annId ? `http://www.animenewsnetwork.com/encyclopedia/anime.php?id=${metadata.annId}` : null,
    }

    return {
      title,
      links,
      studios: metadata.studios ?? null,
      source: metadata.source ?? null,
      schedule: metadata.schedules ? this.serializeSchedules(metadata.schedules) : {},
      durationMinutes: metadata.durationMinutes ?? null,
    }
  }

  private serializeSchedules(schedules: {[country: string]: Schedule}): {[country: string]: WorkSchedule} {
    const result: {[country: string]: WorkSchedule} = {}
    for (const [country, schedule] of Object.entries(schedules)) {
      result[country] = {
        date: schedule.datePrecision !== 'YEAR_MONTH' && schedule.date ?
          Temporal.PlainDate.from(schedule.date)
            .toZonedDateTime(this.defaultTimeZone).toInstant().epochMilliseconds :
          null,
        date_only: schedule.datePrecision ? schedule.datePrecision === 'DATE' : null,
        broadcasts: schedule.broadcasts ?? null,
      }
    }
    return result
  }

  private namuLink(ref: string): string {
    let page = ref
    let anchor = ''
    const anchorIndex = ref.lastIndexOf('#')
    if (anchorIndex >= 0) {
      page = ref.substring(0, anchorIndex)
      anchor = ref.substring(anchorIndex + 1)
    }
    let url = `https://namu.wiki/w/${encodeURIComponent(page)}`
    if (anchor) {
      url += '#' + encodeURIComponent(anchor)
    }
    return url
  }
}
